lib/azuremetadata/azuremetadata.py

```diff
@@ -15,25 +23,73 @@ def get_all(self):
         return result
 
     def get_instance_data(self):
-        # FIXME this is a stub while the CLI UX is being figured out
-
-        if self._api_version == '2019-08-15':
-            with open('fixtures/metadata-v2019-08-15-multiple-network-if.json', 'r') as json_file:
-                return json.load(json_file)
-        elif self._api_version == '2017-04-02':
-            with open('fixtures/metadata-v2017-04-02.json', 'r') as json_file:
-                return json.load(json_file)
-        else:
-            raise Exception("Unknown API version fixture exception")
+        return self._make_request(
+            "http://169.254.169.254/metadata/instance?api-version={}".format(quote(self._api_version))
+        )
 
     def get_attested_data(self):
-        # FIXME this is a stub while the CLI UX is being figured out
-
-        if self._api_version == '2019-08-15':
-            with open('fixtures/attested-data-v2019-08-15.json', 'r') as json_file:
-                return json.load(json_file)
-        elif self._api_version == '2018-10-01':
-            with open('fixtures/attested-data-v2018-10-01.json', 'r') as json_file:
-                return json.load(json_file)
-        else:
-            raise Exception("Unknown API version fixture exception (attested data)")
+        return self._make_request(
+            "http://169.254.169.254/metadata/attested/document?api-version={}".format(quote(self._api_version))
+        )
+
+    def get_disk_tag(self, device=None):
+        if not device:
+            device = self._find_root_device()
+
+        if not device:
+            return ''
+
+        try:
+            with open(device, 'rb') as fh:
+                fh.seek(65536)
+                return str(uuid.UUID(bytes_le=fh.read(16)))
```

Comment:
Please don't use str(). There is no string without an encoding. If you need the string as a string it should be the result of decode(). At best you don't need it at all and handle bytes as long as they are bytes. So you can go with

```
return format(...)
```

or 

```
return ...decode()
```

You can leave the default encoding and trust the system to do it right but str implicitly assumes ascii which might be wrong leading to strange results

Reply:
> You can leave the default encoding and trust the system to do it right but str implicitly assumes ascii which might be wrong leading to strange results

I really doubt. Why would `ascii` be assumed in Python 3? Python 3 strings are supposed to be unicode strings. :thinking: In Python 2 -- maybe.

Which encoding does `format` assume? It does also assume some encoding implicitly (as you aren't supplying it explicitly).

Also, those aren't bytes -- that's an instance of `uuid.UUID`. `str(uuid)` should just call `uuid.__str__()`, which returns a nicely formatted UUID, delimited with dashes.

Reply:
I don't think it is much of an issue in Python3 anymore:

```
Python 2.7.14 (default, Oct 12 2017, 15:50:02) [GCC] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> bytes = b'\xcd\xa1\xc2\xb0 \xcd\x9c\xca\x96 \xcd\xa1\xc2\xb0'
>>> bytes.decode('utf-8') + str(bytes)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xcd in position 0: ordinal not in range(128)
```

```
Python 3.6.5 (default, Mar 31 2018, 19:45:04) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> bytes = b'\xcd\xa1\xc2\xb0 \xcd\x9c\xca\x96 \xcd\xa1\xc2\xb0'
>>> bytes.decode('utf-8') + str(bytes)
"͡° ͜ʖ ͡°b'\\xcd\\xa1\\xc2\\xb0 \\xcd\\x9c\\xca\\x96 \\xcd\\xa1\\xc2\\xb0'"
```

`ascii` is not implicitly assumed anymore and fear of `str` seems irrational to me.